File: challenge-361/perl/ch-1.pl
#!/usr/bin/perl
use v5.38;
use warnings;

use List::Util qw(reduce);

say join(',', zeckendorf($ARGV[0])->@*);

# Return the Zeckendorf representation of $n as an arrayref of Fibonacci numbers
# in descending order, using the greedy algorithm.
sub zeckendorf {
    my ($n) = @_;

    # Build the Fibonacci sequence up to $n using an unfold from [1, 2].
    my @fibs = reverse grep { $_ <= $n }
        map { $_->[0] } generate([1, 2], sub { [$_[0][1], $_[0][0] + $_[0][1]] }, $n);

    # Greedily subtract the largest fitting Fibonacci number until we reach 0,
    # threading a [$remaining, \@result] accumulator tuple through the fold.
    my $acc = reduce {
        my ($remaining, $result) = @$a;
        $b <= $remaining
            ? [$remaining - $b, [@$result, $b]]
            : $a;
    } [$n, []], @fibs;

    return $acc->[1];
}

# Unfold a sequence from a seed tuple by repeatedly applying a generator function.
# Stops when the first element of the current tuple exceeds $limit.
# This is the dual of reduce() (fold). Where reduce() consumes a sequence and
# collapses it into a single value, generate() expands a seed out into a sequence.
sub generate {
    my ($seed, $func, $limit) = @_;
    my @seq;
    my $current = $seed;
    while ($current->[0] <= $limit) {
        push @seq, $current;
        $current = $func->($current);
    }
    return @seq;
}

__END__

File: challenge-361/python/ch-1.py
import sys
from functools import reduce
from itertools import takewhile


def generate(seed, func):
    """Unfold an infinite sequence from a seed by repeatedly applying func.

    This is the dual of reduce() (fold/aggregate). Where reduce() consumes
    a sequence and collapses it into a single value (Iterable[T] -> T),
    generate() expands a single seed value out into a sequence (T -> Iterable[T]).
    Python provides reduce() built-in (via functools), but has no built-in unfold,
    which is why this function is needed.
    """
    current = seed
    while True:
        yield current
        current = func(current)


def zeckendorf(n):
    # Build the Fibonacci sequence up to n using an unfold from (1, 2).
    fibs = list(reversed([
        a for a, b in takewhile(lambda t: t[0] <= n, generate((1, 2), lambda t: (t[1], t[0] + t[1])))
    ]))

    # Greedily subtract the largest fitting Fibonacci number until we reach 0.
    _, result = reduce(
        lambda acc, fib: (acc[0] - fib, acc[1] + [fib]) if fib <= acc[0] else acc,
        fibs,
        (n, [])
    )
    return result


n = int(sys.argv[1])
print(','.join(str(f) for f in zeckendorf(n)))
